src/wymeditor/editor/native-edit-registration.js

```diff
@@ -19,20 +19,49 @@ WYMeditor.NativeEditRegistration = function (wym) {
     // https:/github.com/mightyiam/edited
     nativeEditRegistration.edited = new WYMeditor.EXTERNAL_MODULES.Edited(
         wym._doc.body,
-        function () {
-            nativeEditRegistration._onNativeEdit.call(nativeEditRegistration);
-        }
+        nativeEditRegistration._onSensibleNativeEdit
+            .bind(nativeEditRegistration),
+        nativeEditRegistration._onAnyNativeEdit
+            .bind(nativeEditRegistration)
     );
 };
 
 /**
-    WYMeditor.NativeEditRegistration._onNativeEdit
-    ==============================================
+    WYMeditor.NativeEditRegistration._onSensibleNativeEdit
+    ======================================================
 
-    Handles native edits for the purpose of Undo/Redo.
+    Handles "sensible" native edits.
+    Sensible according to https://github.com/PolicyStat/edited
 */
-WYMeditor.NativeEditRegistration.prototype._onNativeEdit = function () {
-    var nativeEditRegistration = this;
+WYMeditor.NativeEditRegistration.prototype
+    ._onSensibleNativeEdit = function () {
 
+    var nativeEditRegistration = this;
     nativeEditRegistration.wym.registerModification(true);
 };
+
+/**
+    WYMeditor.NativeEditRegistration._onAnyNativeEdit
+    =================================================
+
+    Handles all native edits.
+*/
+WYMeditor.NativeEditRegistration.prototype._onAnyNativeEdit = function () {
+    var nativeEditRegistration = this,
+        undoRedo = nativeEditRegistration.wym.undoRedo;
+
+    // remove redo points
+    undoRedo.history.changesetsFore = [];
+
+    // Non-native modifications are registered when they are performed.
+    // Contrary to those, only a part of the native edits are
+    // registered when they are performed.
+    // A state with unregistered modifications will be lost upon an undo
+    // and redo because it was never registered.
+    // The last registered state doesn't include it and that will be the result
+    // of the undo and redo operation.
+    // This flag lets the undo operation know that it should add the current
+    // state as a history point (register it) before undoing.
+    // Then the next redo will revert to this state.
```

Comment:
You don't have to guess at the "why". You just have to explain it and attempt to simulate the state of yourself after not seeing this code for 12 months. Some folks have trouble with that, which is where code review comes in.

This is a pretty good comment block, but the thing that future-us won't understand from reading this is that native edits are only registered on effectively word/sentence boundaries.

My edit:

Contrary to those, native edits (e.g. typing) aren't registered until word boundaries are reached (space, enter) or edit actions (backspace, delete, cut, paste, etc). This handles an undo after e.g. only part of word is typed.